Symlink fix

tkruse opened the pull request:
(This is an attempt to create a pull request from a fork non-master branch to the main master branch.)

The logic for creating CMakeLists.txt still seems to be flawed. I wrote a unit test for init_workspace and noticed the docstring of init_workspace specified something different from the latest implementation.

So the attached patch changes the logic to:
when linking to toplevel.cmake within workspace use relative path.
Else use absolute path.

Also there was a problem if a user created a workspace with catkin somewhere, using an installed catkin version instead of the catkin version in his workspace. In that case, the init_workspace function would create an absolute link to the installed catkin, not a relative link to the workspace one.

[The more I see this, the less I think it is a good idea to ever allow catkin to live in the same workspace it is used in, as this will always cause non-future-proof hacks.]

Anyway, the patch also detects catkin in the dest workspace via the line in the patch:
workspaces = ['current:' + workspace_dir]

BTW: The whole CATKIN_WORKSPACES logic does not seem to be documented anywhere, I just assumed from the code that this is what is planned.

tkruse commented:
hm, this did not quite work as expected on github

wjwwood commented:
Does it contain commits that you did not intend it to?  You don't have to close it, just update the branch and push -f it, the pull request will update accordingly.

Commit:
code style: comment as sphinx docstring

Commit:
code style: renamed function

Commit:
code style: report causes for error

Commit:
code style: avoid complex one-liners

Commit:
code style: only one function exit at end of function.
This is often regarded better, mostly for refactoring and readability
(because it is easy to forget one of the returns)

Commit:
fix #167: cleaned up init_workspace and reworked rules for relative vs absolute paths. Also simple discovery of catkin within new workspace.

tkruse commented:
Updated to latest master, also this fixes bugs/vulnerabilities in init_workspace around the .CATKIN_WORKSPACE file read

tkruse commented:
the "code style" commits should all change nothing to functionality and can be merged even if the last commit is rejected.

dirk-thomas commented:
I have cherry-picked most of the modifications and made some modifications to the init_workspace part. Please try to keep future pull request as small as possible. I.e. non-essential changes like renaming functions make it more difficult to review.
